bookwyrm/utils/sealed_date.py

```diff
@@ -0,0 +1,240 @@
+
+    def to_python(self, value: Any) -> Optional[SealedDate]:
+        try:
+            date = super().to_python(value)
+        except ValidationError as ex:
+            if match := SelectDateWidget.date_re.match(value):
+                year, month, day = map(int, match.groups())
+            if not match or (day and not month) or not year:
+                raise ex from None
+            if not month:
+                return YearSeal.from_date_parts(year, 1, 1)
+            elif not day:
+                return MonthSeal.from_date_parts(year, month, 1)
+        else:
+            if date is None:
+                return None
+            else:
+                year, month, day = date.year, date.month, date.day
+
+        return SealedDate.from_date_parts(year, month, day)
+
+
+# For typing field and descriptor, below.
+_SetType = datetime
+_GetType = Optional[SealedDate]
+
+
+class SealedDateDescriptor:
+    """descriptor for SealedDateField.
+
+    Encapsulates the "two columns, one field" for SealedDateField.
+    """
+
+    _SEAL_TYPES: dict[Type[_SetType], str] = {
+        YearSeal: "YEAR",
+        MonthSeal: "MONTH",
+        SealedDate: "DAY",
+    }
+
+    _DATE_CLASSES: dict[Any, Type[SealedDate]] = {
+        "YEAR": YearSeal,
+        "MONTH": MonthSeal,
+    }
+
+    def __init__(self, field: models.Field[_SetType, _GetType]):
+        self.field = field
+
+    def __get__(self, instance: models.Model, cls: Any = None) -> _GetType:
+        if instance is None:
+            return self
+
+        value = instance.__dict__.get(self.field.attname)
+
+        if not value or isinstance(value, SealedDate):
+            return value
+
+        # use precision field to construct SealedDate.
+        seal_type = getattr(instance, self.precision_field, None)
+        date_class = self._DATE_CLASSES.get(seal_type, SealedDate)
+
+        return date_class.from_datetime(value)  # FIXME: drop datetimes.
+
+    def __set__(self, instance: models.Model, value: _SetType) -> None:
+        """assign value, with precision where available"""
+        try:
+            seal_type = self._SEAL_TYPES[value.__class__]
+        except KeyError:
+            value = self.field.to_python(value)
+        else:
+            setattr(instance, self.precision_field, seal_type)
+
+        instance.__dict__[self.field.attname] = value
+
+    @classmethod
+    def make_precision_name(cls, date_attr_name: str) -> str:
+        """derive the precision field name from main attr name"""
+        return f"{date_attr_name}_precision"
+
+    @property
+    def precision_field(self) -> str:
+        """the name of the accompanying precision field"""
+        return self.make_precision_name(self.field.attname)
+
+    @property
+    def precision_choices(self) -> list[tuple[str, str]]:
+        """valid options for precision database field"""
+        return [("DAY", "Day seal"), ("MONTH", "Month seal"), ("YEAR", "Year seal")]
```

Comment:
Do these values ever show up in the UI? If so, they should be localized, but I can't find then in the UI anywhere so maybe it doesn't matter

Reply:
They don't ever appear, but it seems Django wants `choices` to always specify the "display version".

(I preferred not to use `("DAY", "DAY"), ...` and made up dummy names instead because can complicate debugging in some unlucky case.)